tstime/tstime.go

```diff
@@ -8,109 +8,119 @@ package tstime
+	if entry, ok := locCache.Load(key); ok {
+		// We're keying only on a hash; double-check zone to ensure no spurious collisions.
+		e := entry.(locCacheEntry)
+		if zone.EqualString(e.zone) {
+			return e.loc, nil
+		}
 	}
 	// TODO(bradfitz): just parse it and call time.FixedLocation.
 	// For now, just have time.Parse do it once:
-	t, err := time.Parse(time.RFC3339Nano, timeValue)
+	t, err := time.Parse(time.RFC3339Nano, timeValue.StringCopy())
 	if err != nil {
 		return nil, err
 	}
 	loc := t.Location()
-	locCache.LoadOrStore(zone, loc)
+	locCache.LoadOrStore(key, locCacheEntry{zone: zone.StringCopy(), loc: loc})
 	return loc, nil
-
 }
 
-// Parse3339 is a wrapper around time.Parse(time.RFC3339Nano, s) that caches
-// timezone Locations for future parses.
-func Parse3339(s string) (time.Time, error) {
+func parse3339m(s mem.RO) (time.Time, error) {
 	zone := zoneOf(s)
-	if zone == "" {
+	if zone.Len() == 0 {
 		// Invalid or weird timezone offset. Use slow path,
 		// which'll probably return an error.
-		return time.Parse(time.RFC3339Nano, s)
+		return time.Parse(time.RFC3339Nano, s.StringCopy())
 	}
 	loc, err := getLocation(zone, s)
 	if err != nil {
 		return time.Time{}, err
 	}
-	s = s[:len(s)-len(zone)] // remove zone suffix
+	s = s.SliceTo(s.Len() - zone.Len()) // remove zone suffix
 	var year, mon, day, hr, min, sec, nsec int
 	const baseLen = len("2020-04-05T15:56:00")
-	if len(s) < baseLen ||
-		!parseInt(s[:4], &year) ||
-		s[4] != '-' ||
-		!parseInt(s[5:7], &mon) ||
-		s[7] != '-' ||
-		!parseInt(s[8:10], &day) ||
-		s[10] != 'T' ||
-		!parseInt(s[11:13], &hr) ||
-		s[13] != ':' ||
-		!parseInt(s[14:16], &min) ||
-		s[16] != ':' ||
-		!parseInt(s[17:19], &sec) {
+	if s.Len() < baseLen ||
+		!parseInt(s.SliceTo(4), &year) ||
+		s.At(4) != '-' ||
+		!parseInt(s.Slice(5, 7), &mon) ||
+		s.At(7) != '-' ||
+		!parseInt(s.Slice(8, 10), &day) ||
+		s.At(10) != 'T' ||
+		!parseInt(s.Slice(11, 13), &hr) ||
+		s.At(13) != ':' ||
+		!parseInt(s.Slice(14, 16), &min) ||
+		s.At(16) != ':' ||
+		!parseInt(s.Slice(17, 19), &sec) {
 		return time.Time{}, errors.New("invalid time")
 	}
-	nsStr := s[baseLen:]
-	if nsStr != "" {
-		if nsStr[0] != '.' {
+	nsStr := s.SliceFrom(baseLen)
+	if nsStr.Len() != 0 {
+		if nsStr.At(0) != '.' {
 			return time.Time{}, errors.New("invalid optional nanosecond prefix")
 		}
-		if !parseInt(nsStr[1:], &nsec) {
-			return time.Time{}, fmt.Errorf("invalid optional nanosecond number %q", nsStr[1:])
+		nsStr = nsStr.SliceFrom(1)
+		if !parseInt(nsStr, &nsec) {
+			return time.Time{}, fmt.Errorf("invalid optional nanosecond number %q", nsStr.StringCopy())
 		}
-		for i := 0; i < len("999999999")-(len(nsStr)-1); i++ {
+		for i := 0; i < len("999999999")-nsStr.Len(); i++ {
 			nsec *= 10
 		}
 	}
 	return time.Date(year, time.Month(mon), day, hr, min, sec, nsec, loc), nil
 }
 
-func parseInt(s string, dst *int) bool {
-	if len(s) == 0 || len(s) > len("999999999") {
+func parseInt(s mem.RO, dst *int) bool {
```

Comment:
the mem package already has:

https://godoc.org/go4.org/mem#ParseInt
https://godoc.org/go4.org/mem#ParseUint

use one of those here now instead? this func can then be a small wrapper

Reply:
That causes a pretty serious slowdown:

```
Parse3339/Z-8              74.3ns ± 1%   121.8ns ± 1%   +63.84%  (p=0.008 n=5+5)
Parse3339/TZ-8              138ns ± 1%     195ns ± 1%   +41.59%  (p=0.008 n=5+5)
ParseInt-8                 8.22ns ± 2%   19.72ns ± 1%  +139.84%  (p=0.008 n=5+5)
```

(For the record, my new function:

```go
func parseInt(s mem.RO, dst *int) bool {
	n, err := mem.ParseUint(s, 10, 64)
	if err != nil {
		return false
	}
	if uint64(int(n)) != n {
		return false
	}
	*dst = int(n)
	return true
}
```

Diffstat is -11 + 5...going to leave it as is.